apps/mobile/src/components/browser/approver-sheet/utils.ts

```diff
@@ -0,0 +1,88 @@
+import {
+  deriveBitcoinAccountFromAddress,
+  useBitcoinAccounts,
+} from '@/store/keychains/bitcoin/bitcoin-keychains.read';
+import { makeAccountIdentifer } from '@/store/utils';
+import { bytesToHex } from '@noble/hashes/utils';
+
+import {
+  getBitcoinInputAddress,
+  getBtcSignerLibNetworkConfigByMode,
+  getPsbtAsTransaction,
+  getPsbtTxInputs,
+  inferPaymentTypeFromAddress,
+  payerToBip32Derivation,
+  payerToTapBip32Derivation,
+} from '@leather.io/bitcoin';
+import {
+  extractAccountIndexFromDescriptor,
+  extractFingerprintFromDescriptor,
+} from '@leather.io/crypto';
+import { BitcoinNetworkModes } from '@leather.io/models';
+import { RpcRequests } from '@leather.io/rpc';
+
+export function normalizeSignAtIndex(signAtIndex: number | number[] | undefined) {
+  if (Array.isArray(signAtIndex)) {
+    return signAtIndex;
+  }
+  if (signAtIndex === undefined) {
+    return undefined;
+  }
+  return [signAtIndex];
+}
+
+interface InfuseBip32Props {
+  psbtHex: string;
+  networkMode: BitcoinNetworkModes;
+  bitcoinAccounts: ReturnType<typeof useBitcoinAccounts>['list'];
+  accountId: string;
+  signAtIndex: number | number[] | undefined;
+}
+
+export function infuseBip32({
+  psbtHex,
+  networkMode,
+  bitcoinAccounts,
+  accountId,
+  signAtIndex: _signAtIndex,
+}: InfuseBip32Props) {
+  const signAtIndex = normalizeSignAtIndex(_signAtIndex) ?? [];
+  const tx = getPsbtAsTransaction(psbtHex);
+
+  const inputs = getPsbtTxInputs(tx);
+  inputs.forEach((input, idx) => {
+    if (input.bip32Derivation || input.tapBip32Derivation) return;
+    const bitcoinNetwork = getBtcSignerLibNetworkConfigByMode(networkMode);
+    const addr = getBitcoinInputAddress(input, bitcoinNetwork);
+    const paymentType = inferPaymentTypeFromAddress(addr);
+    const bitcoinAccount = deriveBitcoinAccountFromAddress(bitcoinAccounts, addr);
```

Comment:
I wonder if this step is necessary. Here, we're looking through _all_ the accounts a wallet has, looking to see if the address matches. `deriveBitcoinAccountFromAddress` also calls the derive fn which can be expensive with many accounts.

We know from the connection history of the app, that it only knows the keys of a given fingerprint/accountId. Can we not then assume that we need to use the account it connected with?